File: ui/goth/assets/src/js/controllers/menu.js
// gothMenu controller (GOTH-1.4; overlay mechanics GOTH-4.1; menu primitives
// GOTH-4.4).
//
// Backs the action-menu primitives — Dropdown Menu (P41), Context Menu (P38), and
// Menubar (P43) — from the shared overlay mechanics: anchored placement of the
// menu panel, roving focus + typeahead over its items, submenu hierarchy, and
// nested-aware Escape/outside dismissal — dispatching the documented
// goth:open / goth:close / goth:select events. Feature-specific event names are
// forbidden (README §8). No new controller name is introduced for GOTH-4.4; this
// refines gothMenu's per-primitive semantics (in-phase), adding context-menu
// pointer/keyboard opening and menubar horizontal coordination.
//
// Structure it discovers: an optional data-slot="trigger" button (dropdown/
// menubar) OR a data-slot="context-trigger" region (context menu), a
// data-slot="content" floating panel (falling back to the root when absent), and
// data-slot="item" items within the panel. A data-slot="submenu-root" wrapping a
// data-slot="submenu-trigger" + data-slot="submenu" panel becomes a submenu. A
// menu whose root is inside a data-slot="menubar" coordinates horizontal roving
// and adjacent-menu opening with its siblings.

import { createRovingFocus } from "../mechanics/roving.js";
import { createDismisser } from "../mechanics/dismiss.js";
import { createTypeahead } from "../mechanics/typeahead.js";
import { createAnchor } from "../mechanics/anchor.js";
import { createSubmenu } from "../mechanics/submenu.js";

function isRTL(el) {
  return (
    el &&
    typeof getComputedStyle === "function" &&
    getComputedStyle(el).direction === "rtl"
  );
}

export default function gothMenu() {
  return {
    open: false,
    _root: null,
    _trigger: null,
    _contextTrigger: null,
    _content: null,
    _items: [],
    _roving: null,
    _typeahead: null,
    _dismisser: null,
    _anchor: null,
    _pointAnchor: null,
    _menubar: null,
    _menubarTriggers: [],

    init() {
      // Cache the root at init; a method invoked from a descendant's x-on handler
      // sees $el bound to that descendant (the GOTH-1.5 descendant-$el fix applied
      // here at source per the GOTH-3.4 Phase-4 follow-up).
      this._root = this.$el;
      this._trigger = this._root.querySelector('[data-slot="trigger"]');
      this._contextTrigger = this._root.querySelector('[data-slot="context-trigger"]');
      this._content = this._root.querySelector('[data-slot="content"]') || this._root;
      // Top-level roving items: the menu items AND submenu triggers, excluding
      // anything nested inside a submenu panel (a submenu owns its own roving).
      this._items = Array.from(
        this._content.querySelectorAll('[data-slot="item"],[data-slot="submenu-trigger"]'),
      ).filter((el) => !el.closest('[data-slot="submenu"]'));
      this._roving = createRovingFocus(this._items, { orientation: "vertical" });
      this._typeahead = createTypeahead((buffer) => {
        const match = this._items.find((el) =>
          (el.textContent || "").trim().toLowerCase().startsWith(buffer),
        );
        if (match) this._roving.setActive(this._items.indexOf(match));
      });
      // Wire each submenu once (its own open/close/roving/anchor lifecycle).
      this._root.querySelectorAll('[data-slot="submenu-root"]').forEach((sub) => {
        const trigger = sub.querySelector('[data-slot="submenu-trigger"]');
        const panel = sub.querySelector('[data-slot="submenu"]');
        if (trigger && panel) createSubmenu(sub, trigger, panel);
      });

      // Menubar coordination: a menu inside a data-slot="menubar" shares a single
      // horizontal tab stop across the sibling triggers and opens the adjacent
      // menu when navigating while one is open.
      this._menubar = this._root.closest('[data-slot="menubar"]');
      if (this._menubar && this._trigger) {
        this._menubar._gothMenus = this._menubar._gothMenus || [];
        this._menubar._gothMenus.push(this);
        this._menubarTriggers = Array.from(
          this._menubar.querySelectorAll('[data-slot="trigger"]'),
        );
        const index = this._menubarTriggers.indexOf(this._trigger);
        this._trigger.setAttribute("tabindex", index === 0 ? "0" : "-1");
        this._trigger.addEventListener("focus", () => this._setMenubarTabstop());
      }
    },

    _focusRestoreTarget() {
      return this._trigger || this._contextTrigger;
    },

    show() {
      if (this.open) return;
      this.open = true;
      this._root.setAttribute("data-state", "open");
      this._content.setAttribute("data-state", "open");
      if (this._trigger) this._trigger.setAttribute("aria-expanded", "true");
      this._roving.init();
      // Escape / selection return focus to the trigger; an outside click leaves
      // focus where the pointer landed.
      this._dismisser = createDismisser(this._root, (reason) => this.hide(reason !== "outside"));
      if (this._pointAnchor) {
        // Context menu: anchor to the pointer position via a virtual zero-size rect.
        const point = this._pointAnchor;
        const virtual = {
          getBoundingClientRect: () => ({
            top: point.y,
            bottom: point.y,
            left: point.x,
            right: point.x,
            width: 0,
            height: 0,
          }),
        };
        this._anchor = createAnchor(virtual, this._content, { side: "bottom", align: "start" });
      } else if (this._trigger && this._content !== this._root) {
        this._anchor = createAnchor(this._trigger, this._content, {
          side: "bottom",
          align: "start",
        });
      }
      this.$nextTick(() => {
        if (this._anchor) this._anchor.activate();
        this._dismisser.activate();
        this._roving.setActive(0);
        this.$dispatch("goth:open");
      });
    },

    hide(refocus = true) {
      if (!this.open) return;
      this.open = false;
      this._root.setAttribute("data-state", "closed");
      this._content.setAttribute("data-state", "closed");
      if (this._trigger) this._trigger.setAttribute("aria-expanded", "false");
      if (refocus) {
        const target = this._focusRestoreTarget();
        if (target) target.focus();
      }
      if (this._dismisser) this._dismisser.deactivate();
      if (this._anchor) {
        this._anchor.deactivate();
        this._anchor = null;
      }
      this._pointAnchor = null;
      this.$dispatch("goth:close");
    },

    toggle() {
      if (this.open) {
        this.hide();
      } else {
        this._pointAnchor = null;
        this.show();
      }
    },

    // openContext opens a context menu at the pointer position (right-click /
    // long-press). It suppresses the browser's native menu.
    openContext(event) {
      event.preventDefault();
      this._pointAnchor = { x: event.clientX, y: event.clientY };
      if (this.open) this.hide(false);
      this.show();
    },

    // onContextKeydown opens a context menu from the keyboard: the ContextMenu key
    // or Shift+F10 (APG) or Enter/Space, anchored at the region's box.
    onContextKeydown(event) {
      const opener =
        event.key === "ContextMenu" ||
        (event.shiftKey && event.key === "F10") ||
        event.key === "Enter" ||
        event.key === " ";
      if (!opener || this.open) return;
      event.preventDefault();
      const region = this._contextTrigger || this._root;
      const box = region.getBoundingClientRect();
      this._pointAnchor = { x: box.left + 8, y: box.top + 8 };
      this.show();
    },

    onTriggerKeydown(event) {
      // ArrowDown/ArrowUp/Enter/Space open the menu (Up focuses the last item).
      if (event.key === "ArrowDown" || event.key === "Enter" || event.key === " ") {
        event.preventDefault();
        if (!this.open) {
          this._pointAnchor = null;
          this.show();
        }
        return;
      }
      if (event.key === "ArrowUp") {
        event.preventDefault();
        if (!this.open) {
          this._pointAnchor = null;
          this.show();
          this.$nextTick(() => this._roving.setActive(this._items.length - 1));
        }
        return;
      }
      // Menubar horizontal navigation across the closed triggers.
      if (this._menubar) {
        const rtl = isRTL(this._trigger);
        const nextKey = rtl ? "ArrowLeft" : "ArrowRight";
        const prevKey = rtl ? "ArrowRight" : "ArrowLeft";
        if (event.key === nextKey) {
          event.preventDefault();
          this._menubarMove(1, false);
        } else if (event.key === prevKey) {
          event.preventDefault();
          this._menubarMove(-1, false);
        }
      }
    },

    onKeydown(event) {
      if (!this.open) return;
      // Menubar: ArrowLeft/ArrowRight (RTL-aware) switch to the adjacent menu and
      // open it. Vertical roving ignores these keys, so they are ours to handle.
      if (this._menubar) {
        const rtl = isRTL(this._content);
        const nextKey = rtl ? "ArrowLeft" : "ArrowRight";
        const prevKey = rtl ? "ArrowRight" : "ArrowLeft";
        if (event.key === nextKey) {
          event.preventDefault();
          this._menubarMove(1, true);
          return;
        }
        if (event.key === prevKey) {
          event.preventDefault();
          this._menubarMove(-1, true);
          return;
        }
      }
      this._roving.onKeydown(event);
      if (event.key && event.key.length === 1) this._typeahead(event.key);
    },

    select(event) {
      const value = event.currentTarget.getAttribute("data-value");
      this.$dispatch("goth:select", { value });
      this.hide();
    },

    // _menubarMove moves focus to the adjacent menu in the bar. step is +1/-1 in
    // logical (direction-adjusted) order; reopen opens the target menu when the
    // current one was open (traversal while a menu is displayed).
    _menubarMove(step, reopen) {
      const triggers = this._menubarTriggers;
      const n = triggers.length;
      if (n === 0) return;
      const i = triggers.indexOf(this._trigger);
      const targetTrigger = triggers[((i + step) % n + n) % n];
      const target = (this._menubar._gothMenus || []).find((m) => m._trigger === targetTrigger);
      const wasOpen = this.open;
      if (wasOpen) this.hide(false);
      if (target) {
        target._focusTrigger();
        if (reopen && wasOpen) target.show();
      } else if (targetTrigger) {
        targetTrigger.focus();
      }
    },

    _focusTrigger() {
      this._setMenubarTabstop();
      if (this._trigger) this._trigger.focus();
    },

    _setMenubarTabstop() {
      this._menubarTriggers.forEach((t) =>
        t.setAttribute("tabindex", t === this._trigger ? "0" : "-1"),
      );
    },
  };
}

File: ui/goth/primitives/menu.go
package primitives

import "github.com/a-h/templ"

// Shared menu helpers for the GOTH-4.4 action-menu primitives: Dropdown Menu
// (P41), Context Menu (P38), and Menubar (P43). All three compose the frozen
// GOTH-4.1 menu mechanics (anchored placement, roving focus + typeahead, submenu
// hierarchy, nested-aware Escape/outside dismissal) through the single gothMenu
// controller — no primitive forks the mechanics and no new controller name is
// introduced (README §8). The div-based floating panel is the same
// .goth-floating layer GOTH-4.1 froze and browser-proved CSP-clean.
//
// No-JS baseline (server-owned, matching the GOTH-4.2 modal/panel precedent): the
// root's server-rendered data-state governs visibility — a menu rendered Open
// shows its content panel with no JavaScript, a closed menu shows only its
// trigger. A pointer/keyboard-invoked action menu has no CSS-only open path, so
// the honest readable baseline is the server-open state (Navigation Menu (P44),
// which is real links, ships a genuinely link-first no-JS baseline instead).
//
// Item roles per the parity rows: plain items are role="menuitem"; the check/
// radio variants carry role="menuitemcheckbox"/role="menuitemradio" with
// aria-checked. State transitions are server-owned (invariant 1): clicking an
// item dispatches goth:select and closes the menu; the app applies the change.

// menuRootAttrs builds the gothMenu root attributes shared by the action menus.
// The server-owned data-state is the no-JS baseline. slot names the primitive's
// stable data-slot (menu / context-menu / menubar-menu).
func menuRootAttrs(b Base, slot string, open bool) templ.Attributes {
	return ownedAttrs(b, templ.Attributes{
		"data-slot":  slot,
		"data-state": openState(open),
		"x-data":     "gothMenu",
	})
}

// menuTriggerAttrs builds a button trigger that toggles the menu on click and
// opens it from the keyboard (ArrowDown/Enter/Space) and, inside a menubar,
// arrows across sibling triggers — all handled by gothMenu.
func menuTriggerAttrs(b Base) templ.Attributes {
	return ownedAttrs(b, templ.Attributes{
		"data-slot":     "trigger",
		"type":          "button",
		"aria-haspopup": "menu",
		"aria-expanded": "false",
		"x-on:click":    "toggle($event)",
		"x-on:keydown":  "onTriggerKeydown($event)",
	})
}

// menuContentAttrs builds the floating menu panel (role=menu). The gothMenu
// controller anchors it under the trigger, roves focus over its items, and runs
// typeahead; keydown is bound here so the panel owns navigation.
func menuContentAttrs(b Base, open bool) templ.Attributes {
	return ownedAttrs(b, templ.Attributes{
		"data-slot":    "content",
		"role":         "menu",
		"data-state":   openState(open),
		"x-on:keydown": "onKeydown($event)",
	})
}

// menuItemAttrs builds a role=menuitem action button. value is the data-value the
// controller reports on goth:select; disabled marks it non-actionable.
func menuItemAttrs(b Base, value string, disabled bool) templ.Attributes {
	owned := templ.Attributes{
		"data-slot":  "item",
		"role":       "menuitem",
		"type":       "button",
		"x-on:click": "select($event)",
	}
	if value != "" {
		owned["data-value"] = value
	}
	if disabled {
		owned["disabled"] = true
		owned["aria-disabled"] = "true"
	}
	return ownedAttrs(b, owned)
}

// menuCheckedItemAttrs builds a role=menuitemcheckbox / role=menuitemradio button
// carrying aria-checked (server-owned state). role is "menuitemcheckbox" or
// "menuitemradio".
func menuCheckedItemAttrs(b Base, role, value string, checked, disabled bool) templ.Attributes {
	owned := templ.Attributes{
		"data-slot":    "item",
		"role":         role,
		"type":         "button",
		"aria-checked": boolString(checked),
		"x-on:click":   "select($event)",
	}
	if value != "" {
		owned["data-value"] = value
	}
	if disabled {
		owned["disabled"] = true
		owned["aria-disabled"] = "true"
	}
	return ownedAttrs(b, owned)
}

// menuLabelAttrs builds a non-interactive group label.
func menuLabelAttrs(b Base) templ.Attributes {
	return ownedAttrs(b, templ.Attributes{"data-slot": "label"})
}

// menuSeparatorAttrs builds a role=separator divider.
func menuSeparatorAttrs(b Base) templ.Attributes {
	return ownedAttrs(b, templ.Attributes{
		"data-slot": "separator",
		"role":      "separator",
	})
}

// menuGroupAttrs builds a role=group wrapper for related items.
func menuGroupAttrs(b Base) templ.Attributes {
	return ownedAttrs(b, templ.Attributes{
		"data-slot": "group",
		"role":      "group",
	})
}

// menuSubRootAttrs builds the submenu wrapper the controller wires as a submenu
// (its own open/close/roving/anchor lifecycle, RTL-aware keys).
func menuSubRootAttrs(b Base) templ.Attributes {
	return ownedAttrs(b, templ.Attributes{"data-slot": "submenu-root"})
}

// menuSubTriggerAttrs builds the submenu trigger (a menuitem that opens the nested
// panel). It carries aria-haspopup="menu" and aria-expanded.
func menuSubTriggerAttrs(b Base) templ.Attributes {
	return ownedAttrs(b, templ.Attributes{
		"data-slot":     "submenu-trigger",
		"type":          "button",
		"role":          "menuitem",
		"aria-haspopup": "menu",
		"aria-expanded": "false",
	})
}

// menuSubContentAttrs builds the nested submenu panel (role=menu).
func menuSubContentAttrs(b Base) templ.Attributes {
	return ownedAttrs(b, templ.Attributes{
		"data-slot":  "submenu",
		"role":       "menu",
		"data-state": "closed",
	})
}

func boolString(b bool) string {
	if b {
		return "true"
	}
	return "false"
}
